File: ios/Runner/EqualizerPlugin.swift
import Foundation
import AVFoundation
import Flutter

public class EqualizerPlugin: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    private var audioEngine: AVAudioEngine?
    private var eqNode: AVAudioUnitEQ?
    private var audioSessionId: UInt32 = 0

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "com.reproductor.equalizer", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "com.reproductor.equalizer/state", binaryMessenger: registrar.messenger())

        let instance = EqualizerPlugin()
        instance.methodChannel = methodChannel
        instance.eventChannel = eventChannel

        methodChannel.setMethodCallHandler(instance.handle(_:result:))
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init":
            guard let args = call.arguments as? [String: Any],
                  let sessionId = args["audioSessionId"] as? NSNumber else {
                result(FlutterError(code: "INVALID_ARGS", message: "audioSessionId required", details: nil))
                return
            }
            initEqualizer(sessionId: sessionId.uint32Value)
            result(nil)

        case "setGain":
            guard let args = call.arguments as? [String: Any],
                  let band = args["band"] as? NSNumber,
                  let gain = args["gain"] as? NSNumber else {
                result(FlutterError(code: "INVALID_ARGS", message: "band and gain required", details: nil))
                return
            }
            setBandGain(band: band.intValue, gain: gain.doubleValue)
            result(nil)

        case "setPreamp":
            guard let args = call.arguments as? [String: Any],
                  let gain = args["gain"] as? NSNumber else {
                result(FlutterError(code: "INVALID_ARGS", message: "gain required", details: nil))
                return
            }
            setPreamp(gain: gain.doubleValue)
            result(nil)

        case "setBassBoost":
            guard let args = call.arguments as? [String: Any],
                  let gain = args["gain"] as? NSNumber else {
                result(FlutterError(code: "INVALID_ARGS", message: "gain required", details: nil))
                return
            }
            // Bass boost is simulated via low-frequency EQ bands
            setBassBoost(gain: gain.doubleValue)
            result(nil)

        case "setVirtualizer":
            guard let args = call.arguments as? [String: Any],
                  let gain = args["gain"] as? NSNumber else {
                result(FlutterError(code: "INVALID_ARGS", message: "gain required", details: nil))
                return
            }
            // Virtualizer not directly available in AVAudioUnitEQ, simulate via EQ
            setVirtualizer(gain: gain.doubleValue)
            result(nil)

        case "setEnabled":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? NSNumber else {
                result(FlutterError(code: "INVALID_ARGS", message: "enabled required", details: nil))
                return
            }
            setEnabled(enabled: enabled.boolValue)
            result(nil)

        case "release":
            release()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initEqualizer(sessionId: UInt32) {
        audioSessionId = sessionId

        audioEngine = AVAudioEngine()
        eqNode = AVAudioUnitEQ(numberOfBands: 10)

        guard let audioEngine = audioEngine, let eqNode = eqNode else { return }

        // Configure 10-band EQ with standard frequencies
        let frequencies: [Float] = [31, 62, 125, 250, 500, 1000, 2000, 4000, 8000, 16000]
        for (index, frequency) in frequencies.enumerated() {
            let band = eqNode.bands[index]
            band.frequency = frequency
            band.filterType = .parametric
            band.bandwidth = 1.0
            band.gain = 0.0
            band.bypass = false
        }

        // Global gain (preamp)
        eqNode.globalGain = 0.0

        audioEngine.attach(eqNode)

        // Connect: player -> eq -> mixer -> output
        // Note: In iOS, we connect to the main mixer
        let format = audioEngine.mainMixerNode.outputFormat(forBus: 0)
        audioEngine.connect(eqNode, to: audioEngine.mainMixerNode, format: format)

        // The audioEngine will be started when the player plays
        // We don't start it here since just_audio manages its own engine

        sendStateUpdate()
    }

    private func setBandGain(band: Int, gain: Double) {
        guard let eqNode = eqNode, band >= 0, band < eqNode.bands.count else { return }
        let clampedGain = max(-12.0, min(12.0, gain))
        eqNode.bands[band].gain = Float(clampedGain)
        sendStateUpdate()
    }

    private func setPreamp(gain: Double) {
        guard let eqNode = eqNode else { return }
        let clampedGain = max(-12.0, min(12.0, gain))
        eqNode.globalGain = Float(clampedGain)
        sendStateUpdate()
    }

    private func setBassBoost(gain: Double) {
        // Boost low frequency bands (0-2: 31Hz, 62Hz, 125Hz)
        guard let eqNode = eqNode else { return }
        let clampedGain = max(0.0, min(12.0, gain))
        for i in 0...2 {
            eqNode.bands[i].gain = Float(clampedGain)
        }
        sendStateUpdate()
    }

    private func setVirtualizer(gain: Double) {
        // Simulate virtualizer by adding slight stereo widening effect via EQ
        // Not directly possible with AVAudioUnitEQ alone, so we just update state
        sendStateUpdate()
    }

    private func setEnabled(enabled: Bool) {
        guard let eqNode = eqNode else { return }
        for band in eqNode.bands {
            band.bypass = !enabled
        }
        eqNode.globalGain = enabled ? eqNode.globalGain : 0.0
        sendStateUpdate()
    }

    private func release() {
        if let eqNode = eqNode {
            audioEngine?.detach(eqNode)
        }
        audioEngine?.stop()
        audioEngine = nil
        eqNode = nil
        sendStateUpdate()
    }

    private func sendStateUpdate() {
        guard let eventSink = eventSink else { return }

        var gains: [Double] = []
        let preamp: Double
        let bassBoost: Double
        let virtualizer: Double
        let enabled: Bool

        if let eqNode = eqNode {
            for band in eqNode.bands {
                gains.append(Double(band.gain))
            }
            preamp = Double(eqNode.globalGain)
            bassBoost = gains.count > 0 ? Double(gains[0]) : 0.0
            virtualizer = 0.0
            enabled = !eqNode.bands[0].bypass
        } else {
            gains = Array(repeating: 0.0, count: 10)
            preamp = 0.0
            bassBoost = 0.0
            virtualizer = 0.0
            enabled = false
        }

        let map: [String: Any] = [
            "gains": gains,
            "preamp": preamp,
            "bassBoost": bassBoost,
            "virtualizer": virtualizer,
            "enabled": enabled
        ]
        eventSink(map)
    }
}

extension EqualizerPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        sendStateUpdate()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: android/app/src/main/kotlin/com/reproductor/reproductor_musica/EqualizerPlugin.kt
package com.reproductor.reproductor_musica

import android.media.audiofx.BassBoost
import android.media.audiofx.Equalizer
import android.media.audiofx.Virtualizer
import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.EventChannel.StreamHandler
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.Executors

class EqualizerPlugin : FlutterPlugin, MethodCallHandler, StreamHandler {
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventSink? = null

    private var equalizer: Equalizer? = null
    private var bassBoost: BassBoost? = null
    private var virtualizer: Virtualizer? = null
    private var audioSessionId: Int = 0
    private var preampValue: Double = 0.0

    private val executor = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "com.reproductor.equalizer")
        methodChannel?.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "com.reproductor.equalizer/state")
        eventChannel?.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        release()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        executor.execute {
            try {
                when (call.method) {
                    "init" -> {
                        val sessionId = call.argument<Int>("audioSessionId") ?: 0
                        initEqualizer(sessionId)
                        mainHandler.post { result.success(null) }
                    }
                    "setGain" -> {
                        val band = call.argument<Int>("band") ?: 0
                        val gain = call.argument<Double>("gain") ?: 0.0
                        setBandGain(band, gain)
                        mainHandler.post { result.success(null) }
                    }
                    "setPreamp" -> {
                        val gain = call.argument<Double>("gain") ?: 0.0
                        setPreamp(gain)
                        mainHandler.post { result.success(null) }
                    }
                    "setBassBoost" -> {
                        val gain = call.argument<Double>("gain") ?: 0.0
                        setBassBoost(gain)
                        mainHandler.post { result.success(null) }
                    }
                    "setVirtualizer" -> {
                        val gain = call.argument<Double>("gain") ?: 0.0
                        setVirtualizer(gain)
                        mainHandler.post { result.success(null) }
                    }
                    "setEnabled" -> {
                        val enabled = call.argument<Boolean>("enabled") ?: false
                        setEnabled(enabled)
                        mainHandler.post { result.success(null) }
                    }
                    "release" -> {
                        release()
                        mainHandler.post { result.success(null) }
                    }
                    else -> mainHandler.post { result.notImplemented() }
                }
            } catch (e: Exception) {
                Log.e("EqualizerPlugin", "Error: ${e.message}")
                mainHandler.post { result.error("ERROR", e.message, null) }
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventSink) {
        eventSink = events
        sendStateUpdate()
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun initEqualizer(sessionId: Int) {
        audioSessionId = sessionId
        try {
            equalizer = Equalizer(0, audioSessionId)
            equalizer?.enabled = true

            bassBoost = BassBoost(0, audioSessionId)
            bassBoost?.enabled = true

            virtualizer = Virtualizer(0, audioSessionId)
            virtualizer?.enabled = true

            sendStateUpdate()
        } catch (e: Exception) {
            Log.e("EqualizerPlugin", "Init error: ${e.message}")
        }
    }

    private fun setBandGain(band: Int, gain: Double) {
        equalizer?.let {
            val count = it.numberOfBands.toInt()
            if (band in 0 until count) {
                val minGain = it.bandLevelRange[0]
                val maxGain = it.bandLevelRange[1]
                val clampedGain = gain.coerceIn(minGain.toDouble() / 100.0, maxGain.toDouble() / 100.0)
                val hundredthsOfDb = (clampedGain * 100).toInt()
                it.setBandLevel(band.toShort(), hundredthsOfDb.toShort())
                sendStateUpdate()
            }
        }
    }

    private fun setPreamp(gain: Double) {
        preampValue = gain
        sendStateUpdate()
    }

    private fun setBassBoost(gain: Double) {
        bassBoost?.let {
            val strength = (gain / 10.0 * 1000).toInt().coerceIn(0, 1000)
            it.setStrength(strength.toShort())
            sendStateUpdate()
        }
    }

    private fun setVirtualizer(gain: Double) {
        virtualizer?.let {
            val strength = (gain / 10.0 * 1000).toInt().coerceIn(0, 1000)
            it.setStrength(strength.toShort())
            sendStateUpdate()
        }
    }

    private fun setEnabled(enabled: Boolean) {
        equalizer?.enabled = enabled
        bassBoost?.enabled = enabled
        virtualizer?.enabled = enabled
        sendStateUpdate()
    }

    private fun release() {
        equalizer?.release()
        equalizer = null
        bassBoost?.release()
        bassBoost = null
        virtualizer?.release()
        virtualizer = null
        sendStateUpdate()
    }

    private fun sendStateUpdate() {
        val sink = eventSink ?: return
        executor.execute {
            val gains = mutableListOf<Double>()
            equalizer?.let { eq ->
                val bandCount = eq.numberOfBands.toInt()
                for (i in 0 until bandCount) {
                    gains.add(eq.getBandLevel(i.toShort()).toDouble() / 100.0)
                }
            }
            val preamp = preampValue
            val bass = bassBoost?.let { it.roundedStrength.toDouble() / 1000.0 * 10.0 } ?: 0.0
            val virtual = virtualizer?.let { it.roundedStrength.toDouble() / 1000.0 * 10.0 } ?: 0.0
            val enabled = equalizer?.enabled ?: false

            val map = mapOf(
                "gains" to gains,
                "preamp" to preamp,
                "bassBoost" to bass,
                "virtualizer" to virtual,
                "enabled" to enabled
            )
            mainHandler.post {
                sink.success(map)
            }
        }
    }
}